From 7e343d3e106b552dded5fc45d85ab0dacb74a60d Mon Sep 17 00:00:00 2001
From: Benjamin Lin <benjamin-jw.lin@mediatek.com>
Date: Thu, 1 Aug 2024 09:11:02 +0800
Subject: [PATCH 174/277] mtk: mt76: mt7996: record per-antenna average
 data-frame RSSI

Record per-antenna average data-frame RSSI.

Signed-off-by: Benjamin Lin <benjamin-jw.lin@mediatek.com>
---
 mt7996/mac.c    | 7 +++++++
 mt7996/main.c   | 5 ++++-
 mt7996/mt7996.h | 2 ++
 3 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index ac51e27d..58b462e8 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -494,10 +494,17 @@ mt7996_mac_fill_rx(struct mt7996_dev *dev, enum mt76_rxq_id q,
 		status->chain_signal[3] = to_rssi(MT_PRXV_RCPI3, v3);
 
 		if (mlink) {
+			int i;
+
 			memcpy(mlink->chain_signal, status->chain_signal,
 			       IEEE80211_MAX_CHAINS);
 			mlink->signal = mt76_rx_signal(mphy->antenna_mask,
 						       mlink->chain_signal);
+
+			for (i = 0; i < IEEE80211_MAX_CHAINS; ++i)
+				ewma_avg_signal_add(mlink->chain_signal_avg + i,
+						    -mlink->chain_signal[i]);
+			ewma_avg_signal_add(&mlink->signal_avg, -mlink->signal);
 		}
 
 		/* RXD Group 5 - C-RXV */
diff --git a/mt7996/main.c b/mt7996/main.c
index f3c55bc5..2dbe4a3b 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1229,7 +1229,7 @@ static int mt7996_add_link_sta(struct mt7996_dev *dev,
 	struct mt7996_sta *msta = (struct mt7996_sta *)sta->drv_priv;
 	u8 link_id = link_sta->link_id;
 	struct mt7996_link_sta *mlink = NULL;
-	int idx, ret;
+	int idx, ret, i;
 
 	if (!rcu_access_pointer(msta->link[link_id])) {
 		idx = mt76_wcid_alloc(dev->mt76.wcid_mask, MT7996_WTBL_STA);
@@ -1297,6 +1297,9 @@ static int mt7996_add_link_sta(struct mt7996_dev *dev,
 			goto error;
 	}
 
+	for (i = 0; i < IEEE80211_MAX_CHAINS; ++i)
+		ewma_avg_signal_init(mlink->chain_signal_avg + i);
+	ewma_avg_signal_init(&mlink->signal_avg);
 	ewma_avg_signal_init(&mlink->avg_ack_signal);
 
 	return 0;
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 38f62d89..59f0be0a 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -321,6 +321,8 @@ struct mt7996_link_sta {
 
 	s8 chain_signal[IEEE80211_MAX_CHAINS];
 	int signal;
+	struct ewma_avg_signal chain_signal_avg[IEEE80211_MAX_CHAINS];
+	struct ewma_avg_signal signal_avg;
 
 	s8 chain_ack_signal[IEEE80211_MAX_CHAINS];
 	int ack_signal;
-- 
2.45.2

